File: BOJ/21278.swift
import Foundation

let INF = 987654321
var mn = INF

var input = readLine()!.split(separator: " ").map{Int(String($0))!}
let n = input[0], m = input[1]

var dist = [[Int]](repeating: [Int](repeating: INF, count: n+1), count: n+1)

for _ in 0..<m {
    input = readLine()!.split(separator: " ").map{Int(String($0))!}
    let a = input[0], b = input[1]
    dist[a][b]=1
    dist[b][a]=1
}

for i in 1...n {
    dist[i][i] = 0
}

var a = 0, b = 0

for k in 1...n {
    for i in 1...n {
        for j in 1...n {
            dist[i][j] = min(dist[i][j], dist[i][k]+dist[k][j])
        }
    }
}

for i in 1...n-1 {
    for j in i+1...n {
        var sum = 0
        for k in 1...n {
            sum += min(dist[i][k] ,dist[j][k])
        }
        if mn > sum {
            mn = sum
            a = i
            b = j
        }
    }
}

print(a,b,2*mn,separator: " ")

File: BOJ/21278.py
import sys

input = sys.stdin.readline
sys.setrecursionlimit(10 ** 6)

INF = 987654321

n, m = map(int, input().split())

adj = [[] for _ in range(n + 1)]
dist = []
mn = INF
a = 0
b = 0
total = [-1]

for i in range(n + 1):
    dist.append([INF for _ in range(n + 1)])
for i in range(1, n + 1):
    dist[i][i] = 0

for _ in range(m):
    a, b = map(int, input().split())
    dist[a][b] = 1
    dist[b][a] = 1

def floyd():
    for k in range(1, n + 1):
        for i in range(1, n + 1):
            for j in range(1, n + 1):
                dist[i][j] = min(dist[i][j], dist[i][k]+dist[k][j])

floyd()

for i in range(1,n+1):
    for j in range(i+1,n+1):
        if i==j: continue
        sum = 0
        for k in range(1,n+1):
            sum += min(dist[i][k], dist[j][k])
        if mn > sum:
            mn = sum
            a = i
            b = j

print(str(a) + " " + str(b) + " " + str(2*mn))
